packages/config/src/files.js

```diff
@@ -1,60 +1,79 @@
 'use strict'
 
-const { resolve } = require('path')
+const { resolve, relative, parse } = require('path')
 
 const { get, set, delete: deleteProp } = require('dot-prop')
 const pathExists = require('path-exists')
 
+const { throwUserError } = require('./error')
 const { mergeConfigs } = require('./merge')
 const { isTruthy } = require('./utils/remove_falsy')
 
 // Make configuration paths relative to `buildDir` and converts them to
 // absolute paths
 const resolveConfigPaths = async function ({ config, repositoryRoot, buildDir, baseRelDir }) {
   const baseRel = baseRelDir ? buildDir : repositoryRoot
-  const configA = resolvePaths(config, FILE_PATH_CONFIG_PROPS, baseRel)
-  const configB = await addDefaultPaths(configA, baseRel)
+  const configA = resolvePaths(config, FILE_PATH_CONFIG_PROPS, baseRel, repositoryRoot)
+  const configB = await addDefaultPaths(configA, repositoryRoot, baseRel)
   return configB
 }
 
 // All file paths in the configuration file are are relative to `buildDir`
 // (if `baseRelDir` is `true`).
 const FILE_PATH_CONFIG_PROPS = ['functionsDirectory', 'build.publish', 'build.edge_handlers']
 
-const resolvePaths = function (config, propNames, baseRel) {
-  return propNames.reduce((configA, propName) => resolvePathProp(configA, propName, baseRel), config)
+const resolvePaths = function (config, propNames, baseRel, repositoryRoot) {
+  return propNames.reduce((configA, propName) => resolvePathProp(configA, propName, baseRel, repositoryRoot), config)
 }
 
-const resolvePathProp = function (config, propName, baseRel) {
+const resolvePathProp = function (config, propName, baseRel, repositoryRoot) {
   const path = get(config, propName)
 
   if (!isTruthy(path)) {
     deleteProp(config, propName)
     return config
   }
 
-  return set(config, propName, resolvePath(baseRel, path))
+  return set(config, propName, resolvePath(repositoryRoot, baseRel, path, propName))
 }
 
-const resolvePath = function (baseRel, path) {
-  if (!isTruthy(path)) {
+const resolvePath = function (repositoryRoot, baseRel, originalPath, propName) {
+  if (!isTruthy(originalPath)) {
     return
   }
 
-  const pathA = path.replace(LEADING_SLASH_REGEXP, '')
-  const pathB = resolve(baseRel, pathA)
-  return pathB
+  const path = originalPath.replace(LEADING_SLASH_REGEXP, '')
+  const pathA = resolve(baseRel, path)
+  validateInsideRoot(originalPath, pathA, repositoryRoot, propName)
+  return pathA
 }
 
 // We allow paths in configuration file to start with /
 // In that case, those are actually relative paths not absolute.
 const LEADING_SLASH_REGEXP = /^\/+/
 
+// We ensure all file paths are within the repository root directory.
+// However we allow file paths to be outside of the build directory, since this
+// can be convenient in monorepo setups.
+const validateInsideRoot = function (originalPath, path, repositoryRoot, propName) {
+  if (relative(repositoryRoot, path).startsWith('..') || getWindowsDrive(repositoryRoot) !== getWindowsDrive(path)) {
```

Comment:
`path.relative()` returns `C:\\...` instead of `..\...` on Windows when the file paths are on different drives.